[process/task] Normalize task types and processes

marechal-p opened the pull request:
It seems like some confusion happened in Theia's implementation of
shell/process tasks, compared to VS Code, but as the goal is to be
compatible with the latter, some cleanup was required.

When executing tasks, we now always use node-pty.

The difference between a process or a shell task relies in the way the
command and its arguments are interpreted/executed:

- process: ask the system to spawn a specific executable, with args
- shell: defer the spawning to a shell that will evaluate a command line

When spawning a process, the expected paramaters are of the kind:

	spawn(executable, [arg1, arg2, arg3, ...])

While when spawning a shell task, we expect some command like

	$> executable arg1 "arg 2" ... && sleep 1 || echo fail

This commit makes Theia tasks and "terminal processes" (processes
spawned via node-pty) able to use both inputs.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

marechal-p commented:
Tried to use tasks when working on C++ things, and this issue bothered me again, hence this draft.

It is heavily inspired from remarks made by @simark regarding the different issues that lied in Theia's Tasks implementation compared to VS Code.

I fixed some tests to adapt to the new behavior, but it shouldn't change very much.

Hopefully it helps you @elaihau with making VS Code extensions able to contribute tasks in Theia, would like to have your feedback on this, see if it improves some things or not.

marechal-p commented on packages/process/src/node/terminal-process.ts:
```diff
@@ -46,6 +52,19 @@ export class TerminalProcess extends Process {
+                options.args = ['/c', command];
+            } else {
+                options.command = 'sh';
+                options.args = ['-i', '-c', command];
```
Might look awkward, [but Node does the same kind of check](https://github.com/nodejs/node/blob/master/lib/child_process.js#L474-L498).

Looking at it maybe we can borrow some things from it?

simark replied:
Eventually, it would be nice to be able to specify which shell to use in the task configuration.  And maybe a preference that says which says to use by default for shell tasks, if unspecified.  Because sh often points to a very basic shell that might not have the features you are used to.

For example, brace expansion.  In bash:

```
$ echo abc{123,456}
abc123 abc456
```

And in dash (which sh points to by default on Debian:

```
$ echo abc{123,456}
abc{123,456}
```

And that's a feature people might want to use in task command lines.

But I think sh is a good default here, since it's the thing that is the most likely to be found everywhere.

The VSCode shell tasks appear to have a key for this:

```
            "options": {
                "shell": {
                    "executable": "...",
                    "args": []
                }
            }
```

marechal-p replied:
Should be able to accept this now :)

Will see how to add a test for it.

simark replied:
If you really want to, I guess you could write your own "shell" (a small node.js script), specify it as the shell executable and somehow verify that it is indeed called with the right arguments.

simark commented on packages/task/src/node/process/process-task-runner.ts:
```diff
@@ -60,9 +58,10 @@ export class ProcessTaskRunner implements TaskRunner {
-        let options;
+        let command: string | undefined;
+        let args: string[] | undefined;
+        let options: any; // tslint:disable-line
```
`// tslint:disable-line:no-any`

simark commented on packages/process/src/node/raw-process.ts:
```diff
@@ -50,31 +51,31 @@ export interface RawProcessFactory {
     readonly process: ChildProcess;
 
+    // shim to not break APIs
+    get outputStream() { return this.output; }
```
Hmm I think it should be the other way around:  outputStream errorStream and inputStream are the "real" fields, and output, input, errorOutput are the getter shims (since the latter are the deprecated ones).

simark replied:
Also, maybe it can be done as a separate cleanup, pushed on its own?

marechal-p replied:
> Also, maybe it can be done as a separate cleanup, pushed on its own?

Are you talking about the addition of streams to the abstract `Process` class?

simark replied:
> Are you talking about the addition of streams to the abstract `Process` class?

Maybe I don't quite understand, but it just looks like a rename from output to outputStream, stdin to stdinStream etc?  If I am mistaken, could you precise why this change is needed?

marechal-p replied:
To normalize the process abstract class, before we had to check which implementation (Process/Terminal) we were using before trying to write/read.

Now since the base class `Process` requires the fields, subclasses just implement it.

I think at some point it helped me with my changes, but with the rewrites maybe it is not directly helpful anymore. I still think it is a good thing.

vince-fugnitto commented on packages/process/src/node/process.ts:
```diff
@@ -78,9 +80,28 @@ export abstract class Process {
     protected _killed = false;
 
+    /**
+     * The OS process id
```
```suggestion
     * The OS process id.
```

vince-fugnitto commented on packages/process/src/node/process.ts:
```diff
@@ -78,9 +80,28 @@ export abstract class Process {
+    abstract readonly pid: number;
+
+    /**
+     * The stdout stream
```
```suggestion
     * The stdout stream.
```

vince-fugnitto commented on packages/process/src/node/process.ts:
```diff
@@ -78,9 +80,28 @@ export abstract class Process {
+    abstract readonly outputStream: Readable;
+
+    /**
+     * The stderr stream
```
```suggestion
     * The stderr stream.
```

vince-fugnitto commented on packages/process/src/node/process.ts:
```diff
@@ -78,9 +80,28 @@ export abstract class Process {
+    abstract readonly errorStream: Readable;
+
+    /**
+     * The stdin stream
```
```suggestion
     * The stdin stream.
```

marechal-p commented on packages/task/src/node/task-server.slow-spec.ts:
```diff
@@ -237,12 +238,16 @@ describe('Task server / back-end', function () {
         const signal = await p;
-        expect(signal).equals('SIGTERM');
+        // expect(signal).equals('SIGTERM');
+        expect(signal).equals('SIGHUP'); // ?
```
Ah, apparently it happened with other test.

My guess is that because when using node-pty, the processes are slaves to some master pty.
So when the slave exits, it would translate to a SIGHUP?

simark replied:
I would venture and say that it's because it's the signal sent by defaut by node-pty when killing a process.

https://github.com/Microsoft/node-pty/blob/master/src/unixTerminal.ts#L226

marechal-p commented:
Where's the CI??

marechal-p commented:
Ah, right when I ask, appveyor shows up, hopefully travis will join too...

marechal-p commented:
@kittaakos hi, sorry to bother you but I think that appveyor is stuck somehow on [this job](https://ci.appveyor.com/project/kittaakos/theia/builds/23155613), you can cancel it if it helps.

kittaakos commented:
> appveyor is stuck somehow

Probably due to this: https://status.appveyor.com/incidents/6wvjng6gr54s

I have restarted the build.

marechal-p commented on packages/task/src/node/task-server.slow-spec.ts:
```diff
@@ -26,26 +26,28 @@ import { terminalsPath } from '@theia/terminal/lib/common/terminal-protocol';
 const bogusCommand = 'thisisnotavalidcommand';
 
-const commandToFindInPathUnix = 'ls';
-const commandToFindInPathWindows = 'dir';
```
Turns out that `dir` is not a real executable, but only exist in a `cmd.exe` prompt...
https://superuser.com/a/229949

marechal-p commented:
All checks are green (except Gitpod, but seems unrelated // cc @kittaakos)!

@elaihau does that LGTY? :)

jankeromnes commented:
> except Gitpod, but seems unrelated

Yes that's unrelated (FYI it's a Gitpod bug that was fixed recently), so your PR is 100% green. 👍

marcdumais-work commented:
> It seems like some confusion happened in Theia's implementation of
shell/process tasks, compared to VS Code, but as the goal is to be
compatible with the latter, some cleanup was required.

With hindsight it can look like that, but I think it's just that @theia/process was done early in the project, and at that time VS Code compatibility was less of a priority. 

In any case it doesn't really matter why, just that we make the requires enhancements if/when needed, like is done in this PR. I would suggest amending the commit message to remove the mention of the "why" and so concentrate on the "what".

marcdumais-work commented on packages/process/src/node/terminal-process.ts:
```diff
@@ -46,6 +52,29 @@ export class TerminalProcess extends Process {
+                options.args = shellArgs || ['/c'];
+                options.args.push(command);
+            } else {
+                options.command = shell || 'sh';
```
"sh" is probably the way to go, but I wonder if some users might expect that the shell being used for shell processes would be the same one being used to spawn terminals (which is configurable, IIRC?

marcdumais-work replied:
nevermind - I think this code already does use the same shell we use for the terminal process.

marcdumais-work commented on packages/task/src/node/process/process-task-runner.ts:
```diff
@@ -89,23 +88,28 @@ export class ProcessTaskRunner implements TaskRunner {
-                    command: command,
-                    args: args,
-                    options: options
+                proc = this.terminalProcessFactory(<TerminalProcessOptions>{
```
Was it voluntary to remove the debug log here but keep the equivalent in the "else" block?

marechal-p replied:
I think I can get rid of both?

Not sure this is useful information?

marechal-p replied:
Or I can add the other one back, will do that.

marcdumais-work replied:
ok to remove both as long as we're consistent

marechal-p commented:
@marcdumais-work :+1:

marcdumais-work commented on packages/task/src/node/task-server.slow-spec.ts:
```diff
@@ -220,29 +224,68 @@ describe('Task server / back-end', function () {
+    /**
+     * TODO: Figure out how to debug a process that correctly starts but exits with a return code > 0
+     */
     it('task using terminal process can handle command that does not exist', async function () {
```
This test fails when I run the unit test  suite on Gitpod; `yarn test:theia` :

13 passing (3s)
  1 failing

  1) Task server / back-end task using terminal process can handle command that does not exist:
     Error: {"taskId":0,"ctx":"/workspace/theia/packages/task/lib/node/../../test-resources","signal":"SIGHUP"}
      at lib/node/task-server.slow-spec.js:399:44
      at /workspace/theia/packages/core/lib/common/event.js:161:33
      at CallbackList.invoke (/workspace/theia/packages/core/lib/common/event.js:176:39)
      at Emitter.fire (/workspace/theia/packages/core/lib/common/event.js:264:29)
      at Object.onTaskExit (lib/common/task-watcher.js:39:29)
      at lib/node/task-server.js:161:20
      at Array.forEach (<anonymous>)
      at TaskServerImpl.fireTaskExitedEvent (lib/node/task-server.js:160:22)
      at lib/node/task-server.js:140:35
      at /workspace/theia/packages/core/lib/common/event.js:161:33
      at CallbackList.invoke (/workspace/theia/packages/core/lib/common/event.js:176:39)
      at Emitter.fire (/workspace/theia/packages/core/lib/common/event.js:264:29)
      at ProcessTask.Task.fireTaskExited (lib/node/task.js:47:26)
      at lib/node/process/process-task.js:59:19
      at /workspace/theia/packages/core/lib/common/event.js:161:33
      at CallbackList.invoke (/workspace/theia/packages/core/lib/common/event.js:176:39)
      at Emitter.fire (/workspace/theia/packages/core/lib/common/event.js:264:29)
      at TerminalProcess.Process.emitOnExit (/workspace/theia/packages/process/lib/node/process.js:123:26)
      at PipeSocket.<anonymous> (/workspace/theia/packages/process/lib/node/terminal-process.js:106:27)
      at UnixTerminal.Terminal.emit (/workspace/theia/node_modules/@theia/node-pty/lib/terminal.js:79:34)
      at PipeSocket.<anonymous> (/workspace/theia/node_modules/@theia/node-pty/lib/unixTerminal.js:85:27)
      at Pipe._handle.close [as _onclose] (net.js:561:12)

marechal-p replied:
It was passing before I added the `shell || process.env["SHELL"] || 'sh'`, no idea why it complains now, will fix!

marechal-p replied:
Ok, so after more investigation, the failure seems to appear at random, that's most likely a race condition.

No real clue why yet.

marechal-p replied:
Ah, I forgot to remove a line that was killing the task (copy/paste from another test).

So it was indeed a race condition due to this line trying to kill the process, vs the task just failing.

marcdumais-work commented:
Testing this PR, I noticed a old bug that it seems has come-back:

Short-running `shell` tasks's output is not displayed in the newly-created terminal widget; instead we get an interactive terminal. I do not see this in current master.

marechal-p commented:
@marcdumais-work just tried on master, and I have the same issue, never saw it fixed on my end...

marcdumais-work commented:
I noticed something else: I think we need to handle the case, where the spawned shell's expansion of the task's command, results in a syntax error. ATM it looks like it's not caught, and the shell process just continues to happily run, processing interactive commands.


e.g., using `<theia repo root>/packages/task/test-resources` as workspace and modifying one of the tasks defined there: 
``` json
        {
            "label": "test task",
            "type": "shell",
            "cwd": "${workspaceFolder}",
            "command": "./task 1\\).. 2\\).. 3)Profit!!!"
        },

```
in `command` one of the closing parenthesis is not escaped, resulting in an error at execution: 

`sh: 1: Syntax error: ")" unexpected`

![Peek 2019-03-21 10-18](https://user-images.githubusercontent.com/25749063/54759205-fbfeda80-4bc3-11e9-97eb-243be55de19e.gif)

marcdumais-work commented:
> just tried on master, and I have the same issue, never saw it fixed on my end...

ah, maybe it's a fine-timing thing? I tried a few times and master behaved correctly for me. Maybe someone else can try?

simark commented:
> I noticed something else: I think we need to handle the case, where the spawned shell's expansion of the task's command, results in a syntax error. ATM it looks like it's not caught, and the shell process just continues to happily run, processing interactive commands.

That's probably because of the `-i` flag we pass to sh.  And the behavior seems different between dash (probably what sh points to on your machine) and bash:

```
zsh$ bash -i -c 'asda ('
bash: syntax error near unexpected token `('
# Does not result in an interactive bash

zsh$ dash -i -c 'asda ('
dash: 1: Syntax error: end of file unexpected (expecting ")")
# Does result in an interact dash
```

Is there a reason to use `-i`?

simark commented:
> ah, maybe it's a fine-timing thing? I tried a few times and master behaved correctly for me. Maybe someone else can try?

I have seen it recently as well. I don't recall that anybody made a conscious effort to fix it, so the underlying problem must still be there.

marcdumais-work commented:
> just tried on master, and I have the same issue, never saw it fixed on my end...

@marechal-p BTW, what did "it" refer-to? The unit test issue or short-lived task's output not displaying in the terminal widget?

marcdumais-work commented:
> I have seen it recently as well. I don't recall that anybody made a conscious effort to fix it, so the underlying problem must still be there.

Confirmed - I just reproduced on master. So it's not a new issue added by this PR.

marechal-p commented:
> Is there a reason to use `-i`?

Not at all, seems to bring more problems as well, will remove :)

vince-fugnitto commented on packages/process/src/node/raw-process.ts:
```diff
@@ -50,30 +51,32 @@ export interface RawProcessFactory {
     (options: RawProcessOptions | RawForkOptions): RawProcess;
 }
 
-/**
```
Since you are refactoring `DevNullStream` into its own file, I think it might be worthwhile to export it from this file as to not break any extenders or users.

marechal-p replied:
done

marechal-p commented:
@elaihau notified me that some VS Code extension contributed tasks were not working with this patch, and this is because I assumed that when running a `shell` task, we would ignore `args` as they should be embedded inside the `command` line.

Turns out VS Code allows you to pass `args`, and they will do the embedding for you inside the command line. I currently do not do that. Looking...

marechal-p commented:
Added the support for the `args` field when spawning terminal processes using a shell.
Added tests alongside it.

Some test failures remain in the tasks, I will start addressing those and it should be all green.

akosyakov commented on packages/process/src/node/raw-process.ts:
```diff
@@ -50,30 +54,32 @@ export interface RawProcessFactory {
+    /**
+     * If the process fails to launch, it will be undefined.
+     */
+    readonly process: ChildProcess | undefined;
```
it is a breaking change

marechal-p replied:
Well, currently the process field can already be undefined, but isn't caught by the type system, so do I revert to the bogus typing?

marechal-p replied:
`process` is assigned here:

https://github.com/theia-ide/theia/blob/bff5375ebbef6228dab924f34c05649b1d790cd5/packages/process/src/node/raw-process.ts#L102-L112

But if the `fork` or `spawn` fail, we cannot assign `process` anymore, leaving it undefined:

https://github.com/theia-ide/theia/blob/bff5375ebbef6228dab924f34c05649b1d790cd5/packages/process/src/node/raw-process.ts#L135-L143

akosyakov replied:
I wonder why in practice it does not cause any issues. Do you know clients using `process`? If  there are not it would be fine to break.

marechal-p replied:
@akosyakov this was the only client in the main sources:

https://github.com/theia-ide/theia/blob/50693c6d380ef3e0aef657550c1f4fd91940df0d/packages/languages/src/node/language-server-contribution.ts#L81

But I think the reason it wasn't failing in practice is because for it to fail, something before has to go wrong, so it just adds one issue on top of another, its like "failing twice". But when things go right this is not an issue.

marechal-p commented:
@elaihau @akosyakov All the checks are green! Although I cheated a little:

I temporarily disabled the tests on Windows for the `Terminal Process { shell: true }` tests. For some reason we get a lot of garbage in the standard output, but it doesn't seem to impact the terminals inside an actual application (everything works).

Once this PR gets merged, I will open an issue marked as `bug` + `help wanted` to fix this issue in the tests, as I really don't know what is causing it.

On the other hand it might have been a bug that was already in, but happened to be silent since the way we started the processes was different before.
